Fixed stripe integration and delete event from google calendar

andreaestefania12 opened the pull request:
## What does this PR do?

- Fixed bug when you trying to book an event with payment, this event was booked and saved in google calendar even if it showed an error. So if you want to book an event with payment, it will show an error and the reservation will not be reserved and will not take availability.

- Fixed bug when you have unpaid or unconfirmed events bookings that need payment and you disconnect the payment method in installed apps, all those bookings are going to be canceled automatically.

Fixes #2170 

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

Test for the first bug.
1. Go to apps/installed.
2. Connect a Stripe account.
3. Create a priced event type.
4. Remove Stripe account.
5. Try to book paid event type.

Test for the second bug.
1. Go to bookings/upcoming and see the unpaid or unconfirmed bookings.
2. Go to apps/installed and disconnect Stripe account.
3. Go to bookings/canceled and you are going to see all unpaid and unconfirmed bookings.

miguelnietoa commented on apps/web/pages/api/book/event.ts:
```diff
@@ -292,16 +292,16 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
+            locale: user.locale ?? "en",
+          },
+        };
+      })
```
I think these changes can be ignored

miguelnietoa commented on apps/web/pages/api/book/event.ts:
```diff
@@ -637,21 +713,23 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
+    // booking successful
+    return res.status(201).json(booking);
+  }
+  return res.status(500).json({ message: "There is not a stripe_Payment credential" });
```
What do you think if we update `stripe_Payment` to `stripe_payment` to be more consistent since the type is like this?

miguelnietoa commented on apps/web/pages/api/integrations.ts:
```diff
@@ -39,12 +49,84 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
+          },
+        });
+
+        await prisma.$transaction([deletePayments, updateBookings, deleteBookingReferences]);
```
I like the idea of using a transaction 👌🏻

miguelnietoa commented on apps/web/pages/api/book/event.ts:
```diff
@@ -507,7 +583,7 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
   try {
     booking = await createBooking();
-    evt.uid = booking.uid;
+    evt.uid = booking ? booking.uid : null;
```
You can refactor this by one of these approaches:
- Optional chaining
  ```js
  evt.uid = booking?.uid;
  ```
  In this case, `evt.uid` will be `undefined` when `booking` is null.

- Null coalescing operator
  ```js
  evt.uid = booking?.uid ?? null;
  ```
  In this case, `evt.uid` will be `null` when `booking` is null.

By using optional chaining, the code will be cleaner but I am not sure if `evt.uid` can be `undefined`.

What do you think?

andreaestefania12 replied:
Hi Miguel, yes the `_evt.uid_` could not be `undefined`.
![image](https://user-images.githubusercontent.com/19562383/162845394-3b3842db-af17-4086-8cdb-e3e0e365d84e.png)

miguelnietoa replied:
Hmm, but the definition type inside the interface `CalendarEvent` specifies that it should be a string or null... So IMHO let's keep it null when there is no booking.

miguelnietoa reviewed (changes requested):
I am testing on my local to check if it works, but I've left some comments.

Let me know if with these changes you have already covered all the use cases for this issue, if yes please mark this PR as **ready for review**.

Good job until now! 💯

miguelnietoa commented:
Hi @andreaestefania12, you also need to **rebase** this branch and resolve conflicts! Let me know if you know how or need assistance!

miguelnietoa commented on apps/web/pages/api/book/confirm.ts:
```diff
@@ -115,6 +116,21 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
       return res.status(400).json({ message: "booking already confirmed" });
     }
 
+    /** If we are confirming a booking that need pay and does not have payment,
+     * it would not have to save in google calendar */
```
Google Calendar is just one of the Calendar integrations, so I think you could replace "google calendar" with "destination calendar" in the comment.

miguelnietoa commented on apps/web/pages/api/integrations.ts:
```diff
@@ -39,12 +49,84 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
     res.status(200).json({ message: "Integration deleted successfully" });
   }
-}
+}
```
Insert a line at the end of this file, please :)

miguelnietoa reviewed (changes requested):
Very good, @andreaestefania12! 👍🏻 
Some minor changes are required.
E2E tests and lint workflows are failing, please review that!

Commit:
Fix merge errors

Commit:
Fix merge errors

Commit:
Errors prettier/prettier

Commit:
Errors prettier/prettier

miguelnietoa commented on apps/web/pages/api/book/event.ts:
```diff
@@ -508,7 +508,27 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
+        id: true,
+      },
+    });
+    /** If the stripe_payment credential exists then it creates the booking OR*/
+    /** If the stripe_payment credential does not exist but the eventTypes is
+    *   without payment then create the booking
+    */
+    if ((eventTypePayment != null && eventType.price > 0) || !eventType.price){
```
```suggestion
    /** If the stripe_payment credential exists then it creates the booking OR*/
    /** If the stripe_payment credential does not exist but the eventTypes is
     *   without payment then create the booking
     */
    if ((eventTypePayment && eventType.price > 0) || !eventType.price) {
```

miguelnietoa commented on apps/web/pages/api/book/event.ts:
```diff
@@ -712,7 +735,12 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
+      eventType.price > 0 &&
+      !originalRescheduledBooking?.paid &&
+      booking
+      ){
```
```suggestion
  if (
    typeof eventType.price === "number" &&
    eventType.price > 0 &&
    !originalRescheduledBooking?.paid &&
    booking
  ) {
```

miguelnietoa commented on apps/web/pages/api/integrations.ts:
```diff
@@ -39,12 +49,84 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
+        await prisma.$transaction([deletePayments, updateBookings, deleteBookingReferences]);
+      } catch (e) {
+        console.error(e);
+        res.status(500).json({ message: "Integration can not be deleted" });
```
```suggestion
        res.status(500).json({ message: "Integration could not be deleted" });
```

miguelnietoa reviewed (changes requested):
Awesome! 💯 

I have only suggested a few small changes related to the code style guides, and let's see if the E2E tests now pass once these changes are implemented.

andreaestefania12 committed:
Update apps/web/pages/api/book/event.ts

Co-authored-by: Miguel Nieto A <39246879+miguelnietoa@users.noreply.github.com>

andreaestefania12 committed:
Update apps/web/pages/api/book/event.ts

Co-authored-by: Miguel Nieto A <39246879+miguelnietoa@users.noreply.github.com>

andreaestefania12 committed:
Update apps/web/pages/api/integrations.ts

Co-authored-by: Miguel Nieto A <39246879+miguelnietoa@users.noreply.github.com>

andreaestefania12 committed:
Merge branch 'calcom:main' into fix/unpaid-unconfirmed

Commit:
NIT changes

andreaestefania12 committed:
Merge branch 'calcom:main' into fix/unpaid-unconfirmed

Commit:
Merge branch 'fix/unpaid-unconfirmed' of github.com:kommitters/cal.com into fix/unpaid-unconfirmed

andreaestefania12 committed:
Update apps/web/pages/api/book/confirm.ts

Co-authored-by: alannnc <alannnc@gmail.com>

Commit:
Modal window before delete stripe integration

Commit:
Merge branch 'fix/unpaid-unconfirmed' of github.com:kommitters/cal.com into fix/unpaid-unconfirmed

Commit:
ESLint Report

andreaestefania12 committed:
Merge branch 'main' into fix/unpaid-unconfirmed

zomars committed:
Merge branch 'main' into pr/2553

zomars committed:
Merge branch 'main' into fix/unpaid-unconfirmed

andreaestefania12 committed:
Merge branch 'main' into fix/unpaid-unconfirmed

andreaestefania12 committed:
Merge branch 'main' into fix/unpaid-unconfirmed

PeerRich committed:
fixed /booking skeleton (#2722)

* fixed /booking skeleton

* nit

andreaestefania12 committed:
Merge branch 'main' into fix/unpaid-unconfirmed

zomars committed:
Type fixes

zomars committed:
Test fixes

zomars committed:
Update playwright.config.ts

zomars committed:
More test fixes

zomars committed:
Update dynamic-booking-pages.test.ts

CarinaWolli committed:
add invite link to Zapier setup page (#2696)

* add invite link and toaster to zapier setup page

* create env variable for invite link and save in database

* fetch invite link form getStaticProps

* add getStaticPath method

* clean code

* Moves app setup and index page

* Moves Loader to ui

* Trying new way to handle dynamic app store pages

* Cleanup

* Update tailwind.config.js

* zapier invite link fixes

* Tests fixes

Co-authored-by: CarinaWolli <wollencarina@gmail.com>
Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>
Co-authored-by: zomars <zomars@me.com>

hariombalhara committed:
Add more embed events (#2719)

* Add more embed events

* Add more embed events

Co-authored-by: Peer Richelsen <peeroke@gmail.com>

alishaz-polymath committed:
adds availability select loader (#2718)

hariombalhara committed:
Improve logs and Fix unwanted 500 to reduce noise in logs (#2674)

* Improve logs

* Fix unintentional 500

Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>

joeauyeung committed:
Change date format for RecurringBookings (#2707)

* Change date format for RecurringBookings

* Missing bookingId query param

Co-authored-by: Leo Giovanetti <hello@leog.me>
Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>

PeerRich committed:
added giphy description (#2730)

PeerRich committed:
fixes #2732 (#2732)

hariombalhara committed:
Hotfix : Fix Infinite loading of Bookings (#2729)

* Add more embed events

* Add more embed events

* Fix nextCursor calculation logic

Co-authored-by: Peer Richelsen <peeroke@gmail.com>

leog committed:
Hotfix: Success page for recurring event (#2725)

* Merge pull request #2672 from calcom/main

v1.5.4

* Turbo fixes

* Make apps single pages public

* Fix preview.html not built and thus served during depooy (#2713)

* Hotfix: Success page layout broken due to duplicate "When" (#2716)

* Update BookingPage.tsx

* Reverting unchaged lines

* Fixing recurrenceRule for ICS files

Co-authored-by: Omar López <zomars@me.com>
Co-authored-by: Hariom Balhara <hariombalhara@gmail.com>

hariombalhara committed:
Fix preview.html not built and thus served during depooy (#2727)

Co-authored-by: Omar López <zomars@me.com>

alishaz-polymath committed:
Allow deletion of a disabled event (#2737)

* allows deletion of disabled event

* some visual fixes

Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>

zomars committed:
Multiple E2E improvements

zomars committed:
Parallelizes some tests

zomars committed:
Update booking-pages.test.ts

zomars committed:
E2E and paid bookings fixes

emrysal committed:
Add 'free' and 'workingElsewhere' as a non-blocking event (#2652)

* Add 'free' and 'workingElsewhere' to non-blocking event - this will allow bookings at these times

* Update CalendarService.ts

Co-authored-by: zomars <zomars@me.com>
Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>

zomars committed:
TODO marks blocking test to be fixed

zomars committed:
Update testUtils.ts

zomars committed:
Update testUtils.ts

zomars committed:
getBusyTimes consolidation

zomars committed:
Fixes delete-me test

zomars committed:
E2E fixing attemps

CarinaWolli committed:
Adjusting Zapier endpoints for publishing integration (#2728)

* add /me endpoint for zapier API testing

* remove cacellationReason from listBookings response

Co-authored-by: CarinaWolli <wollencarina@gmail.com>
Co-authored-by: Peer Richelsen <peeroke@gmail.com>
Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>
Co-authored-by: Omar López <zomars@me.com>

hariombalhara committed:
Fix reschedule not happening in calendar if two calendards are there (#2733)

Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>
Co-authored-by: Peer Richelsen <peeroke@gmail.com>

zomars committed:
Increases timeout temporarily

joeauyeung committed:
Merge pull request #2745 from calcom/apps/multiple-categories

Allow apps to belong to multiple categories

zomars committed:
Build fixes

joeauyeung committed:
Populate msteams key in db (#2743)

* Populate msteams key in db

* Fix calendar credentials to teams

* Clarify account dialog

Co-authored-by: Omar López <zomars@me.com>
Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>
Co-authored-by: Bailey Pumfleet <pumfleet@hey.com>

zomars committed:
Improving Email DRYness (#2486)

* Email DRY

* WIP

* Improve email DRYness

Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>

zomars committed:
E2E fixtures (#2747)

gitstart committed:
Sign in button should be changed or disabled after click #2654 (#2749)

Co-authored-by: gitstart <gitstart@users.noreply.github.com>

leog committed:
Fixing ESLint warnings (#2723)

* Fixing warnings

* Reverting and disabling ESLint in some cases

* Reverting Next Images

* Reverting file, bad merge

* Targeting ESLint to line

* Additional warnings

* New warning squished

* More tweaks and major fixes

* Uneeded conf

Co-authored-by: Omar López <zomars@me.com>

arthur1041 committed:
Fix/avoid multiple schedule deletions (#2602)

* Prevents users from deleting the same schedule multiple times due to delay before the schedule disappears. It also applies the same fix to team disband.

Schedule deletion:
![schedule_deletion_new_behaving](https://user-images.githubusercontent.com/42497300/165126805-b3090268-c1a6-418a-b06e-06bd8446da03.gif)

Team disband:
![team_disband_new_behaving](https://user-images.githubusercontent.com/42497300/165127043-7e083e94-e4c9-4e88-90a2-47d31bdd92e6.gif)

Fixes issue [#2569](https://github.com/calcom/cal.com/issues/2569)

Bug fix (non-breaking change which fixes an issue)

**apps/web/components/LightLoader.tsx** → this file was created in order to make a light color loading spinner available. It's necessary when we need to display a loading spinner above dark backgrounds.

**apps/web/components/availability/ScheduleListItem.tsx** → this component was created in order to give a schedule list item its own state.

* Removing a "setTimeout" that was only used for testing purposes

* Adding a code review suggestion to my modifications

* Changing loading style

* Cleanup

* Avoids using unnecessary state

* Revert "Adding a code review suggestion to my modifications"

This reverts commit b5e40062d71157ca1d9384fb1f5c30d50625809d.

* Reverts some changes

* Renames isLoading

Co-authored-by: Alex van Andel <me@alexvanandel.com>
Co-authored-by: Omar López <zomars@me.com>

Mitchell-Moore committed:
Added check on create eventtype to see is user has slug that already exists (#2757)

* Added check on create eventtype to see is user has slug that already exists, added error check on onError

* revert yarn.lock back

* Catches prisma known error instead of making an additional query

Co-authored-by: zomars <zomars@me.com>

hariombalhara committed:
Fix white border (#2761)

Commit:
fix: remove hardcoded redirect in signin url email verification (#2764)

Co-authored-by: Agusti Fernandez Pardo <git@agusti.me>

zomars committed:
Merge branch 'main' into fix/unpaid-unconfirmed

zomars committed:
Merge branch 'main' into fix/unpaid-unconfirmed

gitstart committed:
Bad UX when user wants to set the default Event Type Title #2245 (#2760)

Co-authored-by: gitstart <gitstart@users.noreply.github.com>

zomars committed:
Skip sending emails in E2E

sean-brydon committed:
Users Phone Number Option (#2669)

* Users Phone Number Option

* Implemented improvments

* Add validation to form

Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>
Co-authored-by: zomars <zomars@me.com>

zomars committed:
Merge branch 'main' into fix/unpaid-unconfirmed

zomars committed:
Test fixes

zomars committed:
Merge branch 'main' into fix/unpaid-unconfirmed

miguelnietoa commented:
Merged into the main repo in https://github.com/calcom/cal.com/pull/2553 🎉
